sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuSubqueryBroadcastExec.scala

```diff
@@ -20,18 +20,124 @@ import scala.collection.JavaConverters.asScalaIteratorConverter
+
+  override def tagPlanForGpu(): Unit = s.child match {
+
+    // For AQE off:
+    //
+    // The rule PlanDynamicPruningFilters will insert SubqueryBroadcast if there exists
+    // available broadcast exchange for reuse. The plan stack of SubqueryBroadcast:
+    //
+    // +- SubqueryBroadcast
+    //    +- BroadcastExchange (can be reused)
+    //       +- [executed subquery...]
+    //
+    // Since the GPU overrides rule has been applied on executedSubQuery, if the
+    // executedSubQuery can be replaced by GPU overrides, the plan stack becomes:
+    //
+    // +- SubqueryBroadcast
+    //    +- BroadcastExchange
+    //       +- GpuColumnarToRow
+    //          +- [GPU overrides of executed subquery...]
+    //
+    // To reuse BroadcastExchange on the GPU, we shall transform above pattern into:
+    //
+    // +- GpuSubqueryBroadcast
+    //    +- GpuBroadcastExchange (can be reused)
+    //       +- [GPU overrides of executed subquery...]
+    //
+    case ex @ BroadcastExchangeExec(_, c2r: GpuColumnarToRowExecParent) =>
+      val exMeta = new GpuBroadcastMeta(ex.copy(child = c2r.child), conf, p, r)
+      exMeta.tagForGpu()
+      if (exMeta.canThisBeReplaced) {
+        broadcastBuilder = () => exMeta.convertToGpu()
+      } else {
+        willNotWorkOnGpu("underlying BroadcastExchange can not run in the GPU.")
+      }
+
+    // For AQE on:
+    //
+    // In Spark 320+, DPP can cooperate with AQE. The insertion of SubqueryBroadcast is
+    // similar with non-AQE circumstance. During the creation of AdaptiveSparkPlan, the
+    // rule PlanAdaptiveSubqueries insert an intermediate plan SubqueryAdaptiveBroadcast to
+    // preserve the initial physical plan of DPP subquery filters. During the optimization,
+    // the rule PlanAdaptiveDynamicPruningFilters inserts the SubqueryBroadcast as the parent
+    // of adaptive subqueries:
+    //
+    // +- SubqueryBroadcast
+    //    +- AdaptiveSparkPlan (supportColumnar=false)
+    //    +- == Initial Plan ==
+    //       BroadcastExchange
+    //       +- [executed subquery...]
+    //
+    // Since AdaptiveSparkPlan can be explicitly set as a columnar plan from Spark 320+,
+    // we can simply build GpuSubqueryBroadcast on the base of columnar adaptive plans
+    // whose root plan are GpuBroadcastExchange:
+    //
+    // +- GpuSubqueryBroadcast
+    //    +- AdaptiveSparkPlan (supportColumnar=true)
+    //    +- == Final Plan ==
+    //       BroadcastQueryStage
+    //       +- GpuBroadcastExchange (can be reused)
+    //          +- [GPU overrides of executed subquery...]
+    //
+    case a: AdaptiveSparkPlanExec =>
+      ShimLoader.getSparkShims.getAdaptiveInputPlan(a) match {
+        case ex: BroadcastExchangeExec =>
+          val exMeta = new GpuBroadcastMeta(ex, conf, p, r)
+          exMeta.tagForGpu()
+          if (exMeta.canThisBeReplaced) {
+            broadcastBuilder = () =>
+              ShimLoader.getSparkShims.columnarAdaptivePlan(a, null)
```

Comment:
I am slightly nervous about passing a `null` for the goal here, even though it is technically safe because the goal isn't currently referenced in the shims for Spark 3.2.0 and later. Perhaps we could use a default goal instead?

```suggestion
              ShimLoader.getSparkShims.columnarAdaptivePlan(a, 
                TargetSize(conf.gpuTargetBatchSizeBytes))
```

Reply:
Done.